### DISCUSSION TITLE
mypyc questions

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [ichard26/ichard26.github.io#4] [Q&A] [answered]
Hello. Enjoyed reading and learning from the 3-part mypyc and Black posts. I have a few questions:

i. Does mypyc handle external libraries, specifically those built with C like numpy, well? 

ii. If Python code includes existing C-extensions (for certain functions) then does mypyc handle them well?

iii. Is there a place to ask mypyc related questions but is not SO?

**Comment by ichard26** [comment, marked as answer]
Hey thanks for reading my blog series, I appreciate the kind words! 

> i. Does mypyc handle external libraries, specifically those built with C like numpy, well?

Probably...? Black uses [typed-ast](https://github.com/python/typed_ast) to parse Python code into an AST on Python 3.7 and below (IIRC?) which is a CPython C Extension and I haven't encountered any weird issues calling to non-mypyc extension code from mypyc code and vice versa yet. In general, I'd assume things should be fine since mypyc would use standard C-APIs to interact with the objects, functions, and classes from external libraries (regardless if it's in C or Python).

> ii. If Python code includes existing C-extensions (for certain functions) then does mypyc handle them well?

I have no idea. I haven't worked with C-extensions before (other than the ones mypyc generate obviously :P). I'm currently learning C so I'd like to write my own C-extensions eventually, but until then I really just don't know. 

> iii. Is there a place to ask mypyc related questions but is not SO?

I have connections to some of the folks who know / work on mypyc so I haven't looked for a place yet, but I'd assume https://gitter.im/python/mypy is your best bet, looks pretty dead though :(

**Reply by dineshbvadhia** [reply]
Thank-you for getting back and with straight-forward answers.

Would you be kind enough to ask the mypyc folks you know about question ii) ?

**Reply by hauntsaninja** [reply]
i. Yes, mypyc handles external libraries just fine, including C extensions. Note that mypyc won't attempt to compile third party libraries you use, it will only attempt to compile what you ask it to compile.

ii. Yes, using C extensions works fine.

iii. https://gitter.im/python/mypy is a place you could ask questions, but there's a lot more chatter about mypy than mypyc over there. You could also consider asking questions at https://github.com/mypyc/mypyc/issues (people are very friendly, but it may take a long while for you to get a response)

**Reply by dineshbvadhia** [reply]
Thank-you so much and greatly appreciated particularly answer to ii). 

In general, is it fair to assume an inline non-mypyc C-extension will be more performant than an mypyc C-extension?

I ask mypy questions at https://github.com/python/typing/discussions/categories/q-a. Is it appropriate for mypyc? Otherwise, will try the gitter channel.

**Reply by hauntsaninja** [reply]
Probably fair to assume that, but it depends on the details.

python/typing discussions is probably not appropriate... You're better off at https://github.com/mypyc/mypyc/issues , like I said :-)
